cmd/kubeadm/app/cmd/init.go

```diff
@@ -502,12 +502,16 @@ func (d *initData) OutputWriter() io.Writer {
 
 // getDryRunClient creates a fake client that answers some GET calls in order to be able to do the full init flow in dry-run mode.
 func getDryRunClient(d *initData) (clientset.Interface, error) {
-	svcSubnetCIDR, err := kubeadmconstants.GetKubernetesServiceCIDR(d.cfg.Networking.ServiceSubnet)
-	if err != nil {
-		return nil, errors.Wrapf(err, "unable to get internal Kubernetes Service IP from the given service CIDR (%s)", d.cfg.Networking.ServiceSubnet)
+	dryRun := apiclient.NewDryRun()
+	if err := dryRun.WithKubeConfigFile(d.KubeConfigPath()); err != nil {
```

Comment:
In dry-run mode, `d.KubeConfigPath()` always returns the kubeconfig path from a kubeadm tmp dir. 

I'm not sure whether the created client is able to talk with the real cluster? 

Do sub phases under init phase need a real cluster when a user run a sub phase with the `--dry-run` flag.

Reply:
@carlory so i approached this differently. the problem is that a tmp dir is used for many files not only the admin.conf.

so i think the only way to allow running individual phases in dry-run mode is to allow the user to specify the dry-run directory. we already had an env variable for that called KUBEADM_INIT_DRYRUN_DIR. i just added new ones for UPGRADE and JOIN as well. so here is how the user can used this.

```
lubo@3B69Q73:~/go/src/k8s.io/kubernetes$ sudo kubeadm init phase certs ca --dry-run
W0925 15:27:34.028290  147394 constants.go:610] Using dry-run directory /etc/kubernetes/tmp/kubeadm-init-dryrun1123083720. To override it, set the environment variable KUBEADM_INIT_DRYRUN_DIR
[certs] Using existing ca certificate authority

lubo@3B69Q73:~/go/src/k8s.io/kubernetes$ sudo ls /etc/kubernetes/tmp/kubeadm-init-dryrun1123083720
ca.crt  ca.key
lubo@3B69Q73:~/go/src/k8s.io/kubernetes$ sudo KUBEADM_INIT_DRYRUN_DIR=/etc/kubernetes/tmp/kubeadm-init-dryrun1123083720
kubeadm init phase kubeconfig all --dry-run
[kubeconfig] Using kubeconfig folder "/etc/kubernetes/tmp/kubeadm-init-dryrun1123083720"
[kubeconfig] Writing "admin.conf" kubeconfig file
[kubeconfig] Writing "super-admin.conf" kubeconfig file
[kubeconfig] Writing "kubelet.conf" kubeconfig file
[kubeconfig] Writing "controller-manager.conf" kubeconfig file
[kubeconfig] Writing "scheduler.conf" kubeconfig file
```

so when not setting the env var a random dir is generted, but if the env var is set it can be used for phase dry-run.